### PR TITLE
hotfix - line follower kernel issue

### PR DESCRIPTION
**Pull request by RobertLucian**
This solves this issue: RobertLucian/GoPiGo#8.
The initial report made by John is here: https://github.com/raspberrypi/firmware/issues/828#issue-236806524

## Explanation

Also, the alternative package that was used for interfacing with the line follower sensor is only supported in Python 3, which means that if someone wants to read data from it, the modules have to be imported in Python 3.

At the moment, I haven't removed the support for Python2, so anything that uses the line follower sensor in Python 2 (aka Scratch or GoPiGo3) won't work. Therefore, this PR needs further modifications before it's merged.

The only single way of keeping Python 2 around is to delve into this issue with the I2C and create a library for it in C++. Then some python bindings would be required in order to make the library importable in python and there goes the time. I find this solution to be best of them all, but it takes time and I don't think it's justifiable to do this **just** so we can keep Python 2 around.

So, I suggest that we move forward with Python 3. In order to do this, we would need to do the following:

1. Rename this branch to a `feature` instead of a `hotfix` and make it PR to the develop branch - the reason is that this is starting to look like a big mod to me and things may go haywire with the master branch.
2. Remove support for Python 2 and exclusively support Python 3 in this repo.
3. Remove support for Python 2 and exclusively support Python 3 in the GoPiGo3 repo.
4. Make DexterOS support Python 3.

Also, to me, I find it peculiar that a sensor which is used on the GoPiGo3 is kept on the GoPiGo repo - I think this has to change as it makes the GoPiGo3 (and thus DexterOS) completely dependent on the GoPiGo repo and this causes us quite some delays because now lots of things have to change - by the way, this is similar to what we have with the distance sensor @CleoQc .

## Testing Hotfix

In order to test the hotfix, follow the instructions:
1. Install the latest kernel for the rpi: `sudo rpi-update 5d03351413d4021cc3feeb04459eea98bcb731c0` - this will update to version `v.4.9.53-v7`.
2. Git checkout my branch.
3. Run the install script found in `/Setup/` folder.
4. Start a python 3 shell and enter:
```
from line_follower import line_sensor
line_sensor.read_sensor()
```

Also, here's a chart with the tests I have run for the GrovePi and the line follower sensor. It looks like the GrovePi works in all cases, which is great, but the line follower sensor doesn't.

kernel_tests

## TL;DR

Can't hotfix kernel issue without pushing for Python 3 and "independence" of the software product. But even so, you should test it just as it is right now.

## Update

@johnisanerd , @CleoQc the python scripts now use the `python-periphery` package as a tool for interfacing with the line follower sensor. I have tested it with the latest kernel version and with Python 2 and Python 3 and it works.

## Merge Requirements

- [x] Add support for installing python packages on DexterOS so we can accept this PR which uses the alternative `python-periphery` package for interfacing with the line follower sensor.

### CODE REVIEW

**Comment by RobertLucian** [comment]
@johnisanerd , @CleoQc the python scripts now use the `python-periphery` package as a tool for interfacing with the line follower sensor. I have tested it with the latest kernel version and with Python 2 and Python 3 and it works.

I say this hotfix is a go.

File: Software/Python/line_follower/line_sensor.py [hunk 1]
```diff
@@ -34,33 +34,22 @@
...
-
-import smbus
 import time
 import math
 import RPi.GPIO as GPIO
 import struct
 import operator
 import pickle
 import numpy
+from periphery import I2C, I2CError
```
Comment (johnisanerd) [inline Software/Python/line_follower/line_sensor.py]: Small request here.  Can you put a comment explaining why we switched to this, and put a link into this PR since you wrote a detailed explanation about what you found (https://github.com/DexterInd/GoPiGo/pull/273/) and maybe also link to the original firmware bug report (https://github.com/raspberrypi/firmware/issues/828) so if we ever have to pick this issue apart again, we have this info a little more readily accessible.

File: Setup/install.sh [hunk 2]
```diff
@@ -81,6 +81,12 @@ install_dependencies() {
     sudo pip install -U RPi.GPIO
     sudo pip install pyusb
     sudo pip install numpy
+    sudo pip install python-periphery
+    sudo pip3 install -U RPi.GPIO
+    sudo pip3 install pyusb
+    sudo pip3 install numpy
+    sudo pip3 install python-periphery
```
Comment (johnisanerd) [inline Setup/install.sh]: An Idea:  clone this repo to the DexterInd account.  Check on changes once per quarter.  This way, changes in the python-periphery library won't suddenly break things for us, forcing us to scramble and fix them.  

Just an idea, not a requirement.

File: Software/Python/line_follower/line_sensor.py (hunk 1 above)
Comment (RobertLucian) [reply Software/Python/line_follower/line_sensor.py]: Done.

File: Setup/install.sh (hunk 2 above)
Comment (RobertLucian) [reply Setup/install.sh]: We're doing the fixed-version thing instead.

File: Setup/install.sh [hunk 3]
```diff
@@ -81,6 +81,12 @@ install_dependencies() {
     sudo pip install -U RPi.GPIO
     sudo pip install pyusb
     sudo pip install numpy
+    sudo pip install python-periphery==1.1.0
```
Comment (johnisanerd) [inline Setup/install.sh]: NICE!

File: Software/Python/line_follower/line_sensor.py [hunk 4]
```diff
@@ -44,23 +42,25 @@
...
+# the line follower no longer works. The issue is with the line follower,
+# which doesn't know how to process repeated starts from the master.
+# With [python-periphery] package, it's possible to avoid issuing
+# repeated start conditions and instead just use start/stop conditions.
+# original report: https://github.com/raspberrypi/firmware/issues/828
+# dev issue      : https://github.com/RobertLucian/GoPiGo/issues/8
+# PR             : https://github.com/DexterInd/GoPiGo/pull/273
+
+
+debug = 0
```
Comment (johnisanerd) [inline Software/Python/line_follower/line_sensor.py]: NICE.  Thank you very much @RobertLucian !

File: Software/Python/requirements.txt [hunk 5]
```diff
@@ -1 +1,2 @@
 future
+quick2wire-api
```
Comment (CleoQc) [inline Software/Python/requirements.txt]: is quick2wire still required?

Comment (RobertLucian) [reply Software/Python/requirements.txt]: Nope. Good catch. This is just a left-over. I'm deleting it.

**Comment by CleoQc** [comment]
Ready to merge?

**Comment by RobertLucian** [comment]
I'm testing it on DexterOS just to be sure nothing gets messed up. If everything looks good, then I'm going to merge it.

**Comment by RobertLucian** [comment]
@CleoQc @johnisanerd tested it on DexterOS and the line follower sensor works with the new library. I've verified the functionality both in the terminal (by importing the sensor and calling the function) and by going in Bloxter and selecting the line follower sensor.
